dagshub/data_engine/model/datasource.py

```diff
@@ -1064,3 +1195,119 @@ def wrap_bytes(val: bytes) -> str:
 
     def get_metadata_entries(self):
         return self._metadata_entries
+
+
+@dataclass
+class DatasourceQuery(DataClassJsonMixin):
+    as_of: Optional[int] = field(default=None, metadata=config(exclude=exclude_if_none, letter_case=LetterCase.CAMEL))
+    select: Optional[List] = field(default=None, metadata=config(exclude=exclude_if_none))
+    filter: "QueryFilterTree" = field(
+        default=QueryFilterTree(),
+        metadata=config(field_name="query", encoder=QueryFilterTree.serialize, decoder=QueryFilterTree.deserialize),
+    )
+
+    def __deepcopy__(self, memodict={}):
+        other = DatasourceQuery(
+            as_of=self.as_of,
+            filter=self.filter.__deepcopy__(),
+        )
+        if self.select is not None:
+            other.select = self.select.copy()
+        return other
+
+    def compose(
+        self, op: str, other: Optional[Union[str, int, float, "DatasourceQuery", "QueryFilterTree", "Datasource"]]
+    ):
+        other_query: Optional["DatasourceQuery"] = None
+        # Extract the filter tree for composition
+        if isinstance(other, (DatasourceQuery, Datasource)):
+            if type(other) is Datasource:
+                other = other.get_query()
+            other_query = other
+            other = other.filter
+
+        self.filter.compose(op, other)
+
+        # If composition went successfully, carry over the as_of and select from the other query
+        if other_query is not None:
+            if other_query.select is not None:
+                self.select = other_query.select
+            if other_query.as_of is not None:
+                self.as_of = other_query.as_of
+
+
+@dataclass
+class DatasourceSerializedState(DataClassJsonMixin):
+    """
+    Serialized state of the datasource.
+    This should be enough to recreate the exact copy of the datasource back (with additional requests)
+
+    Also carries additional information that might be useful for the user:
+        - if the state of the datasource at the point of saving differed from the dataset
+        - the timestamp of saving
+        - link to open the datasource on DagsHub
+    """
+
+    repo: str
+    """Repository this datasource is on"""
+    datasource_id: Union[str, int]
+    """ID of the datasource"""
+    datasource_name: str
+    """Name of the datasource"""
+    query: DatasourceQuery
+    """Query at the time of saving"""
+    dataset_id: Optional[Union[str, int]] = None
+    """ID of the assigned dataset"""
+    dataset_name: Optional[str] = None
+    """Name of the assigned dataset"""
+    timestamp: Optional[float] = None
+    """Timestamp of serialization"""
+    modified: Optional[bool] = None
+    """Does the query differ from the query in the assigned dataset"""
+    link: Optional[str] = None
+    """URL to open this datasource on DagsHub"""
```

Comment:
It's OK that this never gets populated right now?

Reply:
Yea, I checked it, it was no big deal that it was empty